// duo-watch/example/build/app/admin/index.js
(function outer(modules, cache, entries){

  /**
   * Global
   */

  var global = (function(){ return this; })();

  /**
   * Require `name`.
   *
   * @param {String} name
   * @param {Boolean} jumped
   * @api public
   */

  function require(name, jumped){
    if (cache[name]) return cache[name].exports;
    if (modules[name]) return call(name, require);
    throw new Error('cannot find module "' + name + '"');
  }

  /**
   * Call module `id` and cache it.
   *
   * @param {Number} id
   * @param {Function} require
   * @return {Function}
   * @api private
   */

  function call(id, require){
    var m = cache[id] = { exports: {} };
    var mod = modules[id];
    var name = mod[2];
    var fn = mod[0];

    fn.call(m.exports, function(req){
      var dep = modules[id][1][req];
      return require(dep ? dep : req);
    }, m, m.exports, outer, modules, cache, entries);

    // expose as `name`.
    if (name) cache[name] = cache[id];

    return cache[id].exports;
  }

  /**
   * Require all entries exposing them on global if needed.
   */

  for (var id in entries) {
    if (entries[id]) {
      global[entries[id]] = require(id);
    } else {
      require(id);
    }
  }

  /**
   * Duo flag.
   */

  require.duo = true;

  /**
   * Expose cache.
   */

  require.cache = cache;

  /**
   * Expose modules
   */

  require.modules = modules;

  /**
   * Return newest require.
   */

   return require;
})({
1: [function(require, module, exports) {
/**
 * Module Dependencies
 */

var infinity = require('component/infinity');
var event = require('component/event');

console.log('admin!!!');

}, {"component/infinity":2,"component/event":3}],
2: [function(require, module, exports) {
/**
 * Module dependencies
 */

var event = require('event');
var query = require('query');
var throttle = require('throttle');
var debounce = require('debounce');
var Emitter = require('emitter');

/**
 * Export `infinity`
 */

module.exports = infinity;

/**
 * Initialize `infinity`
 *
 * @param {Element|Window} container el
 * @return {infinity}
 * @api public
 */

function infinity(el) {
  if (!(this instanceof infinity)) return new infinity(el);
  this.el = el;
  this.isWindow = (el.self == el);
  this.views = [];
  this._margin = 0;
  this.throttle = throttle(this.refresh.bind(this), 200);
  this.debounce = debounce(this.refresh.bind(this), 100);
  event.bind(el, 'scroll', this.throttle);
  event.bind(el, 'scroll', this.debounce);
  event.bind(el, 'resize', this.debounce);
}

/**
 * Mixin `Emitter`
 */

Emitter(infinity.prototype);

/**
 * Add an element. You may pass any number of arguments
 * to be called on the load and unload functions
 *
 * ex. infinity.add(view.el, view)
 *
 * @param {Element} el
 * @param {Mixed, ...} args
 * @return {infinity}
 * @api public
 */

infinity.prototype.add = function(el) {
  var view = {};
  view.el = el;
  view.args = [].slice.call(arguments) || [];
  view.loaded = false;
  this.views.push(view);
  return this;
};

/**
 * Remove an element.
 *
 * ex. infinity.remove(el)
 *
 * @param {Element} el
 * @return {infinity}
 * @api public
 */

infinity.prototype.remove = function(el) {
  for (var i = 0, view; view = this.views[i]; i++) {
    if (el == view.el) {
      this.views.splice(i, 1);
      break;
    }
  }

  return this;
};


/**
 * Get the coordinated of the box
 *
 * @return {Object} coords
 * @api private
 */

infinity.prototype.box = function() {
  var el = this.el;
  var margin = this._margin;
  var box = {};

  if (!this.isWindow) {
    var rect = el.getBoundingClientRect();
    box.top = rect.top;
    box.left = rect.left;
    box.width = rect.width;
    box.height = rect.height;
  } else {
    box.top = 0;
    box.left = 0;
    box.height = el.innerHeight || document.documentElement.clientHeight;
    box.width = el.innerWidth || document.documentElement.clientWidth;
  }

  box.top -= margin;
  box.left -= margin;
  box.width += (margin * 2);
  box.height += (margin * 2);

  return box;
};

/**
 * Add margin to the box
 *
 * @param {Number} margin
 * @return {infinity}
 * @api public
 */

infinity.prototype.margin = function(margin) {
  this._margin = margin;
  return this;
};

/**
 * Is the element in view?
 *
 * @param {Object} view
 * @return {Boolean}
 * @api private
 */

infinity.prototype.visible = function(view) {
  var box = this._box;
  var pos = view.el.getBoundingClientRect();

  // only load if our view has dimensions
  if (0 === pos.width || 0 === pos.height) return false;

  // check viewport if window otherwise view
  return (this.isWindow) ? this.inViewport(pos, box)
                         : this.inView(pos, box);
};

/**
 * Is the element in view?
 *
 * @param {Object} pos
 * @param {Object} box
 * @return {Boolean}
 * @api private
 */

infinity.prototype.inView = function(pos, box) {
  return (
    pos.top < (box.top + box.height) &&
    pos.left < (box.left + box.width) &&
    (pos.top + pos.height) > box.top &&
    (pos.left + pos.width) > box.left
  );
};

/**
 * Is the element in the viewport?
 *
 * TODO: inViewport and inView could probably be consolidated
 * with some better math
 *
 * @param {Object} pos
 * @param {Object} box
 * @return {Boolean}
 * @api private
 */

infinity.prototype.inViewport = function(pos, box) {
  return (
    pos.bottom >= box.top &&
    pos.right >= box.left &&
    pos.top <= box.height &&
    pos.left <= box.width
  );
};

/**
 * Add a load function
 *
 * @param {Function} fn
 * @return {infinity}
 * @api public
 */

infinity.prototype.load = function(fn) {
  this.on('load', fn);
  return this;
};

/**
 * Add an unload function
 *
 * @param {Function} fn
 * @return {infinity}
 * @api public
 */

infinity.prototype.unload = function(fn) {
  this.on('unload', fn);
  return this;
};

/**
 * Refresh, loading and unloading elements.
 *
 * Used internally but may need to be called
 * manually if you are programmatically adjusting
 * elements.
 *
 * @return {infinity}
 * @api public
 */

infinity.prototype.refresh = function() {
  var visibles = [];
  var invisibles = [];

  this._box = this.box();

  // load / unload panes
  //
  // TODO: figure out a smarter way to not loop
  // through all the elements time but maintain
  // flexibility
  for (var i = 0, view; view = this.views[i]; i++) {
    var visible = this.visible(view);
    if (visible && !view.loaded) {
      visibles.push(view);
    } else if (!visible && view.loaded) {
      invisibles.push(view);
    }
  }

  if (visibles.length) {
    this.emit('loading');
    for (var i = 0, view; view = visibles[i]; i++) {
      this.emit.apply(this, ['load'].concat(view.args));
      view.loaded = true;
    }
  }

  if (invisibles.length) {
    this.emit('unloading');
    for (var i = 0, view; view = invisibles[i]; i++) {
      this.emit.apply(this, ['unload'].concat(view.args));
      view.loaded = false;
    }
  }

  return this;
};

/**
 * Unbind events
 *
 * @return {infinity}
 * @api public
 */

infinity.prototype.unbind = function() {
  event.unbind(this.el, 'scroll', this.throttle);
  event.unbind(this.el, 'scroll', this.debounce);
  event.unbind(this.el, 'resize', this.debounce);
  return this;
};

}, {"event":3,"query":4,"throttle":5,"debounce":6,"emitter":7}],
3: [function(require, module, exports) {
var bind = window.addEventListener ? 'addEventListener' : 'attachEvent',
    unbind = window.removeEventListener ? 'removeEventListener' : 'detachEvent',
    prefix = bind !== 'addEventListener' ? 'on' : '';

/**
 * Bind `el` event `type` to `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @return {Function}
 * @api public
 */

exports.bind = function(el, type, fn, capture){
  el[bind](prefix + type, fn, capture || false);
  return fn;
};

/**
 * Unbind `el` event `type`'s callback `fn`.
 *
 * @param {Element} el
 * @param {String} type
 * @param {Function} fn
 * @param {Boolean} capture
 * @return {Function}
 * @api public
 */

exports.unbind = function(el, type, fn, capture){
  el[unbind](prefix + type, fn, capture || false);
  return fn;
};
}, {}],
4: [function(require, module, exports) {
function one(selector, el) {
  return el.querySelector(selector);
}

exports = module.exports = function(selector, el){
  el = el || document;
  return one(selector, el);
};

exports.all = function(selector, el){
  el = el || document;
  return el.querySelectorAll(selector);
};

exports.engine = function(obj){
  if (!obj.one) throw new Error('.one callback required');
  if (!obj.all) throw new Error('.all callback required');
  one = obj.one;
  exports.all = obj.all;
  return exports;
};

}, {}],
5: [function(require, module, exports) {

/**
 * Module exports.
 */

module.exports = throttle;

/**
 * Returns a new function that, when invoked, invokes `func` at most one time per
 * `wait` milliseconds.
 *
 * @param {Function} func The `Function` instance to wrap.
 * @param {Number} wait The minimum number of milliseconds that must elapse in between `func` invokations.
 * @return {Function} A new function that wraps the `func` function passed in.
 * @api public
 */

function throttle (func, wait) {
  var rtn; // return value
  var last = 0; // last invokation timestamp
  return function throttled () {
    var now = new Date().getTime();
    var delta = now - last;
    if (delta >= wait) {
      rtn = func.apply(this, arguments);
      last = now;
    }
    return rtn;
  };
}

}, {}],
6: [function(require, module, exports) {

/**
 * Module dependencies.
 */

var now = require('date-now');

/**
 * Returns a function, that, as long as it continues to be invoked, will not
 * be triggered. The function will be called after it stops being called for
 * N milliseconds. If `immediate` is passed, trigger the function on the
 * leading edge, instead of the trailing.
 *
 * @source underscore.js
 * @see http://unscriptable.com/2009/03/20/debouncing-javascript-methods/
 * @param {Function} function to wrap
 * @param {Number} timeout in ms (`100`)
 * @param {Boolean} whether to execute at the beginning (`false`)
 * @api public
 */

module.exports = function debounce(func, wait, immediate){
  var timeout, args, context, timestamp, result;
  if (null == wait) wait = 100;

  function later() {
    var last = now() - timestamp;

    if (last < wait && last > 0) {
      timeout = setTimeout(later, wait - last);
    } else {
      timeout = null;
      if (!immediate) {
        result = func.apply(context, args);
        if (!timeout) context = args = null;
      }
    }
  };

  return function debounced() {
    context = this;
    args = arguments;
    timestamp = now();
    var callNow = immediate && !timeout;
    if (!timeout) timeout = setTimeout(later, wait);
    if (callNow) {
      result = func.apply(context, args);
      context = args = null;
    }

    return result;
  };
};

}, {"date-now":8}],
8: [function(require, module, exports) {
module.exports = Date.now || now

function now() {
    return new Date().getTime()
}

}, {}],
7: [function(require, module, exports) {

/**
 * Expose `Emitter`.
 */

module.exports = Emitter;

/**
 * Initialize a new `Emitter`.
 *
 * @api public
 */

function Emitter(obj) {
  if (obj) return mixin(obj);
};

/**
 * Mixin the emitter properties.
 *
 * @param {Object} obj
 * @return {Object}
 * @api private
 */

function mixin(obj) {
  for (var key in Emitter.prototype) {
    obj[key] = Emitter.prototype[key];
  }
  return obj;
}

/**
 * Listen on the given `event` with `fn`.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.on =
Emitter.prototype.addEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};
  (this._callbacks[event] = this._callbacks[event] || [])
    .push(fn);
  return this;
};

/**
 * Adds an `event` listener that will be invoked a single
 * time then automatically removed.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.once = function(event, fn){
  var self = this;
  this._callbacks = this._callbacks || {};

  function on() {
    self.off(event, on);
    fn.apply(this, arguments);
  }

  on.fn = fn;
  this.on(event, on);
  return this;
};

/**
 * Remove the given callback for `event` or all
 * registered callbacks.
 *
 * @param {String} event
 * @param {Function} fn
 * @return {Emitter}
 * @api public
 */

Emitter.prototype.off =
Emitter.prototype.removeListener =
Emitter.prototype.removeAllListeners =
Emitter.prototype.removeEventListener = function(event, fn){
  this._callbacks = this._callbacks || {};

  // all
  if (0 == arguments.length) {
    this._callbacks = {};
    return this;
  }

  // specific event
  var callbacks = this._callbacks[event];
  if (!callbacks) return this;

  // remove all handlers
  if (1 == arguments.length) {
    delete this._callbacks[event];
    return this;
  }

  // remove specific handler
  var cb;
  for (var i = 0; i < callbacks.length; i++) {
    cb = callbacks[i];
    if (cb === fn || cb.fn === fn) {
      callbacks.splice(i, 1);
      break;
    }
  }
  return this;
};

/**
 * Emit `event` with the given args.
 *
 * @param {String} event
 * @param {Mixed} ...
 * @return {Emitter}
 */

Emitter.prototype.emit = function(event){
  this._callbacks = this._callbacks || {};
  var args = [].slice.call(arguments, 1)
    , callbacks = this._callbacks[event];

  if (callbacks) {
    callbacks = callbacks.slice(0);
    for (var i = 0, len = callbacks.length; i < len; ++i) {
      callbacks[i].apply(this, args);
    }
  }

  return this;
};

/**
 * Return array of callbacks for `event`.
 *
 * @param {String} event
 * @return {Array}
 * @api public
 */

Emitter.prototype.listeners = function(event){
  this._callbacks = this._callbacks || {};
  return this._callbacks[event] || [];
};

/**
 * Check if this emitter has `event` handlers.
 *
 * @param {String} event
 * @return {Boolean}
 * @api public
 */

Emitter.prototype.hasListeners = function(event){
  return !! this.listeners(event).length;
};

}, {}]}, {}, {"1":""})
